### PR TITLE
add DNS e2e tests

### PR DESCRIPTION
**Pull request by Choraden**
This patch adds: 
- support for defining custom networks in compose tool
- bind9 dns server service to use in e2e tests
- DNS e2e setup

### CODE REVIEW

File: utils/compose/compose.go [hunk 1]
```diff
@@ -16,16 +16,18 @@ import (
 )
 
 type Compose struct {
-	Path     string              `yaml:"-"`
-	Version  string              `yaml:"version"`
-	Services map[string]*Service `yaml:"services,omitempty"`
+	Path     string                    `yaml:"-"`
+	Version  string                    `yaml:"version"`
+	Services map[string]*Service       `yaml:"services,omitempty"`
+	Networks map[string]*NetworkConfig `yaml:"networks,omitempty"`
```
Comment (mmatczuk) [inline utils/compose/compose.go:22]: Why NetworkConfig not Network?

Comment (Choraden) [reply utils/compose/compose.go:22]: Naming conflict between docker network definition and service network. Feel free to suggest something better.

File: utils/compose/network.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+func validateIP(input string) error {
+	if ip := net.ParseIP(input); ip == nil {
+		return fmt.Errorf("could not parse IP address: %s", input)
+	}
+
+	return nil
+}
+
+func validateIPWithMask(input string) error {
+	parts := strings.Split(input, "/")
```
Comment (mmatczuk) [inline utils/compose/network.go:41]: Maybe `Cut`

File: utils/compose/network.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+	if len(parts) != 2 {
+		return fmt.Errorf("must be in the format of IP/MASK, input: %s", input)
+	}
+
+	if ip := net.ParseIP(parts[0]); ip == nil {
```
Comment (mmatczuk) [inline utils/compose/network.go:46]: Maybe call validateIP

File: utils/compose/network.go [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+	}
+
+	mask, err := strconv.Atoi(parts[1])
+	if err != nil {
+		return fmt.Errorf("could not parse mask: %s", parts[1])
+	}
+
+	if mask < 0 || mask > 32 {
+		return fmt.Errorf("must be between 0 and 32, mask: %d", mask)
+	}
```
Comment (mmatczuk) [inline utils/compose/network.go:57]: Refactor to validateMask

File: utils/compose/network.go [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+	for i := range n.IPAM.Config {
+		if err := validateIPWithMask(n.IPAM.Config[i].Subnet); err != nil {
+			return fmt.Errorf("network subnet is invalid: %w", err)
+		}
+		if err := validateIP(n.IPAM.Config[i].Gateway); err != nil {
+			return fmt.Errorf("network gateway is invalid %w", err)
+		}
+		if err := validateIPWithMask(n.IPAM.Config[i].IPRange); err != nil {
+			return fmt.Errorf("network IP range is invalid %w", err)
+		}
```
Comment (mmatczuk) [inline utils/compose/network.go:78]: Refactor to method.

File: utils/compose/service.go [hunk 6]
```diff
@@ -11,14 +11,19 @@ import (
 	"time"
 )
 
+type Network struct {
```
Comment (mmatczuk) [inline utils/compose/service.go:14]: I think this should be renamed to ServiceNetwotk

File: e2e/forwarder/service.go [hunk 7]
```diff
@@ -131,6 +131,13 @@ func (s *Service) WithEnv(key, val string) *Service {
 	return s
 }
 
+func (s *Service) WithNetwork(network, ipv4 string) *Service {
```
Comment (mmatczuk) [inline e2e/forwarder/service.go:135]: How about with IP?

File: e2e/dns/db.rpz [hunk 8]
```diff
@@ -0,0 +1,10 @@
+$TTL 60
+@            IN    SOA  localhost. root.localhost.  (
+                          2015112501   ; serial
+                          1h           ; refresh
+                          30m          ; retry
+                          1w           ; expiry
+                          30m)         ; minimum
+                   IN     NS    localhost.
+
+httpbin.saucelabs.test.com    A        192.168.100.10
```
Comment (mmatczuk) [inline e2e/dns/db.rpz:10]: I'm not a huge fan of this address, I do not like using a subdomain of test.com 

I might suggest httpbin.local?

File: e2e/dns/service.go [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+package dns
+
+import (
+	"github.com/saucelabs/forwarder/utils/compose"
+)
+
+type service compose.Service
+
+const (
+	Image = "ubuntu/bind9:9.18-22.04_beta"
```
Comment (mmatczuk) [inline e2e/dns/service.go:16]: legacy of the future, how about latest here?

File: e2e/forwarder/service.go [hunk 10]
```diff
@@ -138,6 +139,11 @@ func (s *Service) WithNetwork(network, ipv4 string) *Service {
 	return s
 }
 
+func (s *Service) WithDNS(servers ...string) *Service {
```
Comment (mmatczuk) [inline e2e/forwarder/service.go:142]: WithDNSServer - the flag name is DNS_SERVER not DNS

File: e2e/tests/dns_test.go [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+package tests
+
+import (
+	"net/http"
+	"testing"
+)
+
+var httpbinDNS = serviceScheme("HTTPBIN_PROTOCOL") + "://httpbin.saucelabs.test.com:8080"
+
+func TestDNS(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (mmatczuk) [reply e2e/tests/dns_test.go:18]: IMO this is a flag test and should be named TestFlagDNSServer

File: e2e/setups.go [hunk 12]
```diff
@@ -213,3 +217,43 @@ func SC2450Setup() setup.Setup {
 		Run: "^TestSC2450$",
 	}
 }
+
+func DNSSetup() setup.Setup {
```
Comment (mmatczuk) [inline e2e/setups.go:221]: Again this is a flag test, follow conventions for testing flags.

File: e2e/dns/service.go [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+		},
+	}
+}
+
+func (s *service) WithNetwork(network, ipv4 string) *service {
+	s.Network = map[string]compose.Network{
+		network: {IPv4: ipv4},
+	}
+	return s
+}
```
[the remark this answers is not in the document]
Comment (Choraden) [reply e2e/dns/service.go:39]: Currently we don't have any `With` method in compose.Service. We make new type to define our own methods
```go
type Service compose.Service
```
This is why we also have general use method:
```go
func (s *Service) WithEnv(key, val string) *Service {
	s.Environment[key] = val
	return s
}
```
defined in e2e.forwarder package.

[the remark this answers is not in the document]
Comment (Choraden) [reply e2e/dns/service.go:39]: Ugh, `Type '*Service' has both field and method named 'Service'` I see why we did this. 
I'll leave it as it is for now.

**Comment by Choraden** [comment]
@mmatczuk I've addressed your comments.

**Comment by mmatczuk** [comment]
LGTM

**Pull request merged**